R/NimbleAPI.R

```diff
@@ -0,0 +1,133 @@
+#' @include LabKeySettings.R
+#' @include Utils.R
+
+library(dplyr)
+
+DownloadAndAppendNimble <- function(seuratObject, outPath=".") {#, genomes=NULL) {
+  
+  # Ensure we have a DatasetId column
+  if (is.null(seuratObject@meta.data[['DatasetId']])) {
+    stop('Seurat object lacks DatasetId column')
+  }
+  
+  # Produce a nimble file id/DatasetId vector for each DatasetId
+  nimbleFileComponents <- list()
+  for (datasetID in unique(seuratObject@meta.data[['DatasetId']])) {
+    print(paste0('Possibly adding nimble data for dataset: ', datasetID))
+    
+    nimbleIds <- .queryNimble(loupeDataId=datasetID)#, genomeFilter=genomes)
+    
+    # TODO Make this configurable whether or not we skip or return
+    if (is.null(nimbleIds)) {
+      print(paste0('Nimble file(s) not found for dataset: ', datasetID, ', skipping'))
+      next
+    }
+    
+    for(id in nimbleIds) {
+      nimbleFile <- file.path(outPath, paste0(id, '_nimbleCounts.tsv'))
+      
+      DownloadOutputFile(outputFileId = id, outFile = nimbleFile, overwrite = T)
+      
+      if (!file.exists(nimbleFile)) {
+        stop(paste0('Unable to download calls table for nimbleId: ', nimbleId, ' datasetId: ', datasetId))
+      }
+      
+      nimbleFileComponents <- append(nimbleFileComponents, list(c(nimbleFile, datasetID)))
+    }
+  }
+  
+  print(nimbleFileComponents)
+  df <- .mergeNimbleFiles(fileComponents=nimbleFileComponents)
+  
+  outFile <- file.path(outPath, paste0("mergedNimbleCounts.tsv"))
```

Comment:
should this be configurable from an argument? Maybe, "writeOutputTable = TRUE"?

Reply:
Reconciling this with my comment from above about tmpDir(), here's what I suggest:

- eliminate the outDir arg. Always save the raw TSVs to tempdir(). Also be sure to unlink() the file when done.
- Have an argument for rawCountsFile = NULL. If this is null, dont write the table. If non-null, then write the table to that path.

what do you think?

Reply:
That makes sense, especially wrt the tmpdir stuff -- I was kind of wondering why the default was to dump to the local dir, instead.

Reply:
in other contexts it might have been debugging, but this code is also most often exercised headlessly in a docker where that ends up being thrown away. It's probable some of the existing patterns could be better...